fraud/registry.go

```diff
@@ -1,17 +1,45 @@
 package fraud
 
 import (
+	"errors"
+	"fmt"
 	"sync"
 )
 
 var (
 	unmarshalersLk      = sync.RWMutex{}
 	defaultUnmarshalers = map[ProofType]ProofUnmarshaler{}
+
+	proofsLk            = sync.RWMutex{}
+	supportedProofTypes = map[ProofType]string{}
 )
 
-// Register sets unmarshaler in map by provided ProofType.
-func Register(proofType ProofType, unmarshaler ProofUnmarshaler) {
+// Register sets supported proofs with it string representation and unmarshalers in maps by provided ProofType.
+func Register(p Proof) error {
+	proofsLk.Lock()
+	if _, ok := supportedProofTypes[p.Type()]; ok {
+		proofsLk.Unlock()
+		return errors.New("fraud: proofType is registered")
+	}
+	supportedProofTypes[p.Type()] = p.Name()
+	proofsLk.Unlock()
+
 	unmarshalersLk.Lock()
-	defer unmarshalersLk.Unlock()
-	defaultUnmarshalers[proofType] = unmarshaler
+	defaultUnmarshalers[p.Type()] = func(data []byte) (Proof, error) {
+		err := p.UnmarshalBinary(data)
```

Comment:
Hmm, there is an issue here. P is not copied and the same value might be reused causing races. Can we make a copy of an empty proof somehow before unmarshalling it?

Reply:
created a copy of proof